packages/replay-next/src/suspense/SyntaxParsingCache.ts

```diff
@@ -1,407 +1,157 @@
-        // This typically indicates white space or newline characters.
-        processSection(currentLineState, parsedTokens, codeToParse.slice(characterIndex, from), "");
-      }
+      state.totalLength = source.codeUnitCount;
 
-      processSection(currentLineState, parsedTokens, codeToParse.slice(from, to), classes);
+      const chunk = source.contents.slice(processedLength);
 
-      characterIndex = to;
-    });
+      const lines = chunk.split(/\r\n?|\n|\u2028|\u2029/);
+      lines[0] = state.rawText.pop() + lines[0];
+      state.rawText = state.rawText.concat(lines);
+      state.rawProgress = source.contents.length / source.codeUnitCount!;
+      update(state.parsedTokens, state.parsedProgress, {
+        text: state.rawText,
+        progress: state.rawProgress,
+      });
 
-    const maxPosition = codeToParse.length - 1;
-    if (characterIndex < maxPosition) {
-      // No style applied on the trailing text.
-      // This typically indicates white space or newline characters.
-      processSection(
-        currentLineState,
-        parsedTokens,
-        codeToParse.slice(characterIndex, maxPosition),
-        ""
-      );
-    }
+      const message: ParseRequest = {
+        type: "parse",
+        sourceId,
+        chunk,
+        isLastChunk: source.complete,
+      };
+      postWorkerMessage(message);
 
-    if (currentLineState.parsedTokens.length) {
-      parsedTokens.push(currentLineState.parsedTokens);
+      processedLength += chunk.length;
     }
 
-    parsedCharacterIndex += characterIndex + 1;
-
-    complete = isCodeComplete && parsedCharacterIndex >= code.length;
-
-    // Anything that's left should de-opt to plain text.
-    if (parsedCharacterIndex < codeToParse.length) {
-      let nextIndex = codeToParse.indexOf("\n", parsedCharacterIndex);
-
-      let parsedLineTokens: ParsedToken[] = [];
-
-      while (true) {
-        const line =
-          nextIndex >= 0
-            ? codeToParse.substring(parsedCharacterIndex, nextIndex)
-            : codeToParse.substring(parsedCharacterIndex);
-
-        parsedLineTokens.push({
-          columnIndex: 0,
-          types: null,
-          value: line,
-        });
+    source.subscribe(processChunk);
 
-        if (nextIndex >= 0) {
-          parsedTokens.push(parsedLineTokens);
-
-          parsedLineTokens = [];
-        } else if (nextIndex === -1) {
-          break;
-        }
-
-        parsedCharacterIndex = nextIndex + 1;
-        nextIndex = codeToParse.indexOf("\n", parsedCharacterIndex);
-      }
-
-      if (parsedLineTokens.length) {
-        parsedTokens.push(parsedLineTokens);
-      }
+    if (source.lineCount !== null) {
+      processChunk();
     }
-  }
+  },
+});
 
-  return {
-    isComplete: () => complete,
-    parsedTokensByLine: parsedTokens,
-    parseChunk,
-  };
+export function parseStreaming(source: StreamingSourceContents, fileName: string | null) {
+  return streamingParserCache.stream(source, fileName);
```

Comment:
I think we should start exporting these caches _directly_ rather than wrapper methods. The caches usually have multiple methods that are potentially useful. The "suspense" package also exports utils/hooks for working with the caches that can't be used if the actual caches can't be referenced.